zkevm-circuits/src/evm_circuit/execution/calldatasize.rs

```diff
@@ -0,0 +1,179 @@
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+
+use crate::{
+    evm_circuit::{
+        step::ExecutionState,
+        table::CallContextFieldTag,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition},
+            Cell,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+
+use super::ExecutionGadget;
+
+#[derive(Clone, Debug)]
+pub(crate) struct CallDataSizeGadget<F> {
+    same_context: SameContextGadget<F>,
+    call_data_size: Cell<F>,
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for CallDataSizeGadget<F> {
+    const NAME: &'static str = "CALLDATASIZE";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::CALLDATASIZE;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        // Calldatasize can be looked up in the above tx_id's context.
+        let call_data_size = cb.query_cell();
+
+        // Add lookup constraint in the call context for the calldatasize field.
+        cb.call_context_lookup(
+            None,
+            CallContextFieldTag::CallDataLength,
+            call_data_size.expr(),
+        );
+
+        // The calldatasize should be pushed to the top of the stack.
+        cb.stack_push(call_data_size.expr());
```

Comment:
I believe that the CallDataLength from the CallContext is the actual value, but in the stack it must be RLC encoded:
https://github.com/appliedzkp/zkevm-specs/blob/feb9d0e9a42cd6ef3a62ae0d3e87cd9224100bb6/src/zkevm_specs/evm/execution/calldatasize.py#L13-L22

Reply:
@ed255 I've made changes to resolve the above comments and discussions.

Reply:
I think this implementation is still different than the spec.  Currently the value used for the stack WRITE is not RLC encoded.

Take a look for example at the COINBASE implementation: https://github.com/appliedzkp/zkevm-circuits/blob/b1e368ec7812ca7db9a016b780d586e247f634a3/zkevm-circuits/src/evm_circuit/execution/coinbase.rs#L32  The coinbase value is an ethereum address so it fits in the field.  In the RwTable.CallContext it's raw, and then in the RwTable.Stack it's RLC encoded.